libyul/backends/evm/SSAControlFlowGraphBuilder.cpp

```diff
@@ -112,45 +132,45 @@ SSACFG::ValueId tryRemoveTrivialPhi(SSACFG& _cfg, SSACFG::ValueId _phi)
 				if (input == _phi)
 					uses.emplace_back(Use{std::ref(op), n});
 		}
-		block.forEachExit([&](SSACFG::BlockId _block) {
-			_recurse(_block, _recurse);
-		});
-	};
-	auto phiBlock = phiInfo->block;
-	_cfg.block(phiBlock).phis.erase(_phi);
-	findUses(phiBlock, findUses);
-
-	if (!same.hasValue())
-	{
-		// This will happen for unreachable paths.
-		// TODO: check how best to deal with this
-		same = _cfg.unreachableValue();
+		// todo check if it is sound to set "unreachable" return values to literal 0
+		if (auto* functionReturn = std::get_if<SSACFG::BasicBlock::FunctionReturn>(&block.exit))
+			std::replace(
+				functionReturn->returnValues.begin(),
+				functionReturn->returnValues.end(),
+				_phi,
+				same == m_graph.unreachableValue() ? m_graph.newLiteral(block.debugData, 0) : same
+			);
```

Comment:
What would happen if you instead wrote a zero-value to each return variable at the beginning of each function?

Reply:
That should also do the trick and is a bit more idiomatic. I like it!

Reply:
Yeah, less of a headache wondering if it's correct in any case! And I'd assume it'd also work.

Reply:
It's funny, adding an explicit `retval := 0` to the entry block in yul fixes things, doing it via `currentDef(returnVar, entry) = zero()` does not. :D Not sure yet what is going on there.

Reply:
It's funny, adding an explicit `retval := 0` to the entry block in yul fixes things, doing it via `currentDef(returnVar, entry) = zero()` does not. :D Not sure yet what is going on there.

Reply:
Weird, sounds like that should be exactly equivalent :-).

Reply:
That was also my thinking but it seems it's a different problem altogether that just is symptomatically cured by adding an explicit assignment of the return variable into the entry block :)

The deeper symptom is, that it tries to optimize phi functions which have no arguments (which is expected I guess), but for some reason the return value ends up being a value dependent on a phi function defined in the entry block without any args. So now im digging why that may happen:)

![image](https://github.com/user-attachments/assets/714ad677-ffe2-48d9-aa3e-2786e762afaf)